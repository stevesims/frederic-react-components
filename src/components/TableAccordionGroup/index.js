import React from 'react';
import PropTypes from 'prop-types';
import glamorous from 'glamorous';
import { Collapse } from 'react-collapse';
import { FONT_SIZE } from '@govuk-react/constants';

import OpenButton from '../OpenButton';

const Container = glamorous.div({
  display: 'flex',
  fontFamily: 'Roboto, Arial, sans-serif',
  fontSize: FONT_SIZE.SIZE_14,
  lineHeight: '1.75',
  paddingBottom: '10px',
  paddingTop: '10px',
  position: 'relative',
});

const Title = glamorous.header(({ clickable }) => (clickable ? { cursor: 'pointer' } : undefined), {
  paddingLeft: '16px',
});

const Items = glamorous.div({});
const FirstItem = glamorous.div({});

const TableAccordionGroup = ({ title, children, expanded, onChange, open, changeOnTitleClick }) => {
  return (
    <Container>
      <OpenButton open={open} onChange={onChange} />
      <Title clickable={changeOnTitleClick} onClick={changeOnTitleClick && (() => onChange({ open: !open }))}>
        {title}
      </Title>
      <Items>
        <FirstItem>{children}</FirstItem>
        <Collapse isOpened={open}>{expanded}</Collapse>
      </Items>
    </Container>
  );
};

TableAccordionGroup.propTypes = {
  changeOnTitleClick: PropTypes.bool,
  children: PropTypes.node.isRequired,
  expanded: PropTypes.node.isRequired,
  onChange: PropTypes.func,
  open: PropTypes.bool,
  title: PropTypes.node,
};

TableAccordionGroup.defaultProps = {
  changeOnTitleClick: false,
  open: false,
};

export default TableAccordionGroup;

export { Title };
