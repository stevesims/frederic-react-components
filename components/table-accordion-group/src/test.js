import React from 'react';
import { mount, shallow } from 'enzyme';
import { createMatchers } from 'jest-emotion';
import * as emotion from 'emotion';

import TableAccordionGroup from '.';

expect.extend(createMatchers(emotion));

describe('TableAccordionGroup', () => {
  it('renders with only required properties', () => {
    const wrapper = shallow(<TableAccordionGroup expanded="Expanded">Test</TableAccordionGroup>);
    expect(wrapper.exists()).toBe(true);
  });

  it('renders title when passed', () => {
    const wrapper = mount(<TableAccordionGroup expanded="Expanded" title="Title">Test</TableAccordionGroup>);
    expect(
      wrapper
        .find('Title')
        .childAt(0)
        .text(),
    ).toBe('Title');
  });

  it('triggers onChange when OpenButton is clicked', () => {
    const changeHandler = jest.fn();
    const wrapper = mount(
      <TableAccordionGroup expanded="Expanded" onChange={changeHandler}>
        Test
      </TableAccordionGroup>,
    );
    wrapper.find('OpenButton').simulate('click');
    expect(changeHandler).toHaveBeenCalledTimes(1);
  });
  
  it('does not trigger onChange when title is clicked  and changeOnTitleClick is not true', () => {
    const changeHandler = jest.fn();
    const wrapper = mount(
      <TableAccordionGroup expanded="Expanded" onChange={changeHandler}>
        Test
      </TableAccordionGroup>,
    );
    wrapper.find('Title').simulate('click');
    expect(changeHandler).not.toHaveBeenCalled();
  });

  it('triggers onChange when title is clicked and changeOnTitleClick is true', () => {
    const changeHandler = jest.fn();
    const wrapper = mount(
      <TableAccordionGroup expanded="Expanded" onChange={changeHandler} changeOnTitleClick>
        Test
      </TableAccordionGroup>,
    );
    wrapper.find('Title').simulate('click');
    expect(changeHandler).toHaveBeenCalledTimes(1);
  });

  it('matches snapshot', () => {
    const changeHandler = jest.fn();
    const wrapper = mount(
      <TableAccordionGroup expanded="Expanded" onChange={changeHandler} changeOnTitleClick>
        Test
      </TableAccordionGroup>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
