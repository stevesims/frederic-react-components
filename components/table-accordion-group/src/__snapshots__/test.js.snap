// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableAccordionGroup matches snapshot 1`] = `
.emotion-6 {
  margin-bottom: 15px;
}

@media only screen and (min-width:641px) {
  .emotion-6 {
    margin-bottom: 15px;
  }
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: "nta",Arial,sans-serif;
  font-size: 14px;
  line-height: 1.75;
  padding-bottom: 10px;
  padding-top: 10px;
  position: relative;
  margin-bottom: 15px;
}

@media only screen and (min-width:641px) {
  .emotion-4 {
    margin-bottom: 15px;
  }
}

.emotion-0 {
  background: none;
  border: none;
  cursor: pointer;
  height: 20px;
  padding: 0;
  -webkit-transition: -webkit-transform 0.3s ease-in-out;
  -webkit-transition: transform 0.3s ease-in-out;
  transition: transform 0.3s ease-in-out;
  width: 20px;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
}

.emotion-0:focus {
  outline: none;
}

.emotion-2 {
  cursor: pointer;
  padding-left: 16px;
}

<Styled(TableAccordionGroup)
  changeOnTitleClick={true}
  expanded="Expanded"
  onChange={[MockFunction]}
  open={false}
>
  <TableAccordionGroup
    changeOnTitleClick={true}
    className="emotion-6"
    expanded="Expanded"
    onChange={[MockFunction]}
    open={false}
  >
    <Container
      className="emotion-6"
    >
      <div
        className="emotion-4 emotion-5"
      >
        <OpenButton
          onChange={[MockFunction]}
          open={false}
        >
          <RotatingButton
            onClick={[Function]}
            open={false}
          >
            <button
              className="emotion-0 emotion-1"
              onClick={[Function]}
              open={false}
            >
              <Arrow
                fill="#0C60A2"
                width={20}
              >
                <svg
                  height="20px"
                  viewBox="0 0 6 4"
                  width="20px"
                >
                  <polygon
                    fill="#0C60A2"
                    points="3 4 6 1 0 1"
                  />
                </svg>
              </Arrow>
            </button>
          </RotatingButton>
        </OpenButton>
        <Title
          clickable={true}
          onClick={[Function]}
        >
          <header
            className="emotion-2 emotion-3"
            onClick={[Function]}
          />
        </Title>
        <div>
          <div>
            Test
          </div>
          <Collapse
            fixedHeight={-1}
            forceInitialAnimation={false}
            hasNestedCollapse={false}
            isOpened={false}
            onMeasure={[Function]}
            onRender={[Function]}
            onRest={[Function]}
            style={Object {}}
            theme={
              Object {
                "collapse": "ReactCollapse--collapse",
                "content": "ReactCollapse--content",
              }
            }
          >
            <Motion
              defaultStyle={
                Object {
                  "height": 0,
                }
              }
              onRest={[Function]}
              style={
                Object {
                  "height": 0,
                }
              }
            >
              <div
                className="ReactCollapse--collapse"
                style={
                  Object {
                    "height": 0,
                    "overflow": "hidden",
                  }
                }
              >
                <div
                  className="ReactCollapse--content"
                >
                  Expanded
                </div>
              </div>
            </Motion>
          </Collapse>
        </div>
      </div>
    </Container>
  </TableAccordionGroup>
</Styled(TableAccordionGroup)>
`;
